std/sumtype.d

```diff
@@ -0,0 +1,2488 @@
+
+        /**
+         * Handles formatted writing of the `SumType`'s current value.
+         *
+         * Not available when compiled with `-betterC`.
+         *
+         * Params:
+         *   sink = Output range to write to.
+         *   fmt = Format specifier to use.
+         *
+         * See_Also: $(REF formatValue, std,format)
+         */
+        void toString(this This, Sink, Char)(ref Sink sink, const ref FormatSpec!Char fmt);
+    }
+
+    version (D_BetterC) {} else
+    /**
+     * Returns a string representation of the `SumType`'s current value.
+     *
+     * Not available when compiled with `-betterC`.
+     */
+    string toString(this This)()
+    {
+        import std.conv : to;
+
+        return this.match!(to!string);
+    }
+
+    version (D_BetterC) {} else
+    /**
+     * Handles formatted writing of the `SumType`'s current value.
+     *
+     * Not available when compiled with `-betterC`.
+     *
+     * Params:
+     *   sink = Output range to write to.
+     *   fmt = Format specifier to use.
+     *
+     * See_Also: $(REF formatValue, std,format)
+     */
+    void toString(this This, Sink, Char)(ref Sink sink, const ref FormatSpec!Char fmt)
+    {
+        import std.format : formatValue;
+
+        this.match!((ref value) {
+            formatValue(sink, value, fmt);
+        });
+    }
+
+    static if (allSatisfy!(isHashable, Map!(ConstOf, Types)))
+    {
+        // Workaround for dlang issue 21400
+        version (StdDdoc)
+        {
+            /**
+             * Returns the hash of the `SumType`'s current value.
+             *
+             * Not available when compiled with `-betterC`.
+             */
+            size_t toHash() const;
+        }
+
+        // Workaround for dlang issue 20095
+        version (D_BetterC) {} else
+        /**
+         * Returns the hash of the `SumType`'s current value.
+         *
+         * Not available when compiled with `-betterC`.
+         */
+        size_t toHash() const
+        {
+            return this.match!hashOf;
+        }
+    }
+}
+
+// Construction
+@safe unittest
+{
+    alias MySum = SumType!(int, float);
+
+    assert(__traits(compiles, MySum(42)));
+    assert(__traits(compiles, MySum(3.14)));
+}
+
+// Assignment
+@safe unittest
+{
+    alias MySum = SumType!(int, float);
+
+    MySum x = MySum(42);
+
+    assert(__traits(compiles, x = 3.14));
+}
+
+// Self assignment
+@safe unittest
+{
+    alias MySum = SumType!(int, float);
+
+    MySum x = MySum(42);
+    MySum y = MySum(3.14);
+
+    assert(__traits(compiles, y = x));
+}
+
+// Equality
+@safe unittest
+{
+    alias MySum = SumType!(int, float);
+
+    MySum x = MySum(123);
+    MySum y = MySum(123);
+    MySum z = MySum(456);
+    MySum w = MySum(123.0);
+    MySum v = MySum(456.0);
+
+    assert(x == y);
+    assert(x != z);
+    assert(x != w);
+    assert(x != v);
+
+}
+
+// Equality of differently-qualified SumTypes
+version (D_BetterC) {} else
```

Comment:
I'd add a comment explaining why this can't be tested in this version.

Reply:
Done.